cabal-install/Distribution/Client/RebuildMonad.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+-- | An abstraction for re-running actions if values or files have changed.
+--
+-- This is not a full-blown make-style incremental build system, it's a bit
+-- more ad-hoc than that, but it's easier to integrate with existing code.
+--
+-- It's a convenient interface to the "Distribution.Client.FileMonitor"
+-- functions.
+--
+module Distribution.Client.RebuildMonad (
+    -- * Rebuild monad
+    Rebuild,
+    runRebuild,
+
+    -- * Setting up file monitoring
+    monitorFiles,
+    MonitorFilePath(..),
+    monitorFileSearchPath,
+    FilePathGlob(..),
+
+    -- * Using a file monitor
+    FileMonitorCacheFile(..),
+    rerunIfChanged,
+
+    -- * Utils
+    matchFileGlob,
+  ) where
+
+import Distribution.Client.FileStatusCache
+         ( MonitorFilePath(..), monitorFileSearchPath
+         , FilePathGlob(..), matchFileGlob
+         , FileMonitorCacheFile(..), Changed(..)
+         , checkFileMonitorChanged, updateFileMonitor )
+
+import Distribution.Simple.Utils (debug)
+import Distribution.Verbosity    (Verbosity)
+
+import Control.Applicative
+import Control.Monad.State as State
+import Data.Maybe      (fromMaybe)
+import Data.Binary     (Binary)
+import System.FilePath (takeFileName)
+
+
+-------------------------------
+-- Simple rebuild abstraction
+--
+
+newtype Rebuild a = Rebuild (StateT [MonitorFilePath] IO a)
```

Comment:
Oh, very interesting! Maybe we should remark that it is just used as Writer, but it's not Writer to avoid space leak.